libretro/libretro.cpp

```diff
@@ -531,16 +538,26 @@ static void update_input(void)
 {
    // Reset input states
    u32 J = 0;
+   int16_t joy_bits = 0;
+   unsigned i;
 
    /* if (retropad_device[0] == RETRO_DEVICE_JOYPAD) */ {
+      if (libretro_supports_bitmasks)
+         joy_bits = input_cb(0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_MASK);
+      else
+      {
+         for (i = 0; i < MAX_BUTTONS; i++)
+            joy_bits |= input_cb(0, RETRO_DEVICE_JOYPAD, 0, i) ? (1 << i) : 0;
+      }
+
       for (unsigned button = 0; button < MAX_BUTTONS; button++)
-         J |= input_cb(0, RETRO_DEVICE_JOYPAD, 0, binds[button]) << button;
+         J |= joy_bits & (1 << binds[button]) ? (1 << button) : 0;
 
       if (option_turboEnable) {
          /* Handle Turbo A & B buttons */
          bool button_pressed = false;
          for (unsigned tbutton = 0; tbutton < TURBO_BUTTONS; tbutton++) {
-            if (input_cb(0, RETRO_DEVICE_JOYPAD, 0, turbo_binds[tbutton])) {
+            if (joy_bits & (1 << turbo_binds[tbutton])) {
```

Comment:
this will never activate on non-bitmask mode since the indes assigned for dedicated turbo buttons are not part of MAX_BUTTONS count

Reply:
Oh bugger! I see that now. I'll just read in the whole array for future safety.

Reply:
https://github.com/libretro/vba-next/pull/139